tests/e2e/runtime_reconfiguration_test.go

```diff
@@ -0,0 +1,201 @@
+			ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+			defer cancel()
+
+			epc, err := e2e.NewEtcdProcessCluster(ctx, t, e2e.WithClusterSize(tc.clusterSize))
+			require.NoError(t, err)
+			require.NoError(t, epc.Procs[0].Etcdctl().Health(ctx))
+			defer func() {
+				err := epc.Close()
+				require.NoError(t, err, "failed to close etcd cluster: %v", err)
+			}()
+
+			for i := 0; i < 2; i++ {
+				time.Sleep(etcdserver.HealthInterval)
+				if !tc.asLearner {
+					addMember(ctx, t, epc)
+				} else {
+					addMemberAsLearner(ctx, t, epc)
+				}
+			}
+		})
+	}
+}
+
+func TestRuntimeReconfigDecreaseClusterSize(t *testing.T) {
+	e2e.BeforeTest(t)
+
+	tcs := []struct {
+		name        string
+		clusterSize int
+		asLearner   bool
+	}{
+		{
+			name:        "decrease cluster size from 3 to 1",
+			clusterSize: 3,
+		},
+		{
+			name:        "decrease cluster size from 5 to 3",
+			clusterSize: 5,
+		},
+	}
+	for _, tc := range tcs {
+		t.Run(tc.name, func(t *testing.T) {
+			ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+			defer cancel()
+
+			epc, err := e2e.NewEtcdProcessCluster(ctx, t, e2e.WithClusterSize(tc.clusterSize))
+			require.NoError(t, err)
+			require.NoError(t, epc.Procs[0].Etcdctl().Health(ctx))
+			defer func() {
+				err := epc.Close()
+				require.NoError(t, err, "failed to close etcd cluster: %v", err)
+			}()
+
+			for i := 0; i < 2; i++ {
+				time.Sleep(etcdserver.HealthInterval)
+				removeMember(ctx, t, epc)
+			}
+		})
+	}
+}
+
+func TestRuntimeReconfigRollingUpgrade(t *testing.T) {
+	e2e.BeforeTest(t)
+
+	tcs := []struct {
+		name        string
+		withLearner bool
+	}{
+		{
+			name:        "with learner",
+			withLearner: true,
+		},
+		{
+			name:        "without learner",
+			withLearner: false,
+		},
+	}
+
+	for _, tc := range tcs {
+		t.Run(tc.name, func(t *testing.T) {
+			ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+			defer cancel()
+
+			epc, err := e2e.NewEtcdProcessCluster(ctx, t, e2e.WithClusterSize(3))
+			require.NoError(t, err)
+			require.NoError(t, epc.Procs[0].Etcdctl().Health(ctx))
+			defer func() {
+				err := epc.Close()
+				require.NoError(t, err, "failed to close etcd cluster: %v", err)
+			}()
+
+			for i := 0; i < 2; i++ {
+				time.Sleep(etcdserver.HealthInterval)
+				removeMember(ctx, t, epc)
+				epc.WaitLeader(t)
+				// if we do not wait for leader, without the fix of notify raft Advance,
+				// have to wait 1 sec to pass the test stably.
+				if tc.withLearner {
+					addMemberAsLearner(ctx, t, epc)
+				} else {
+					addMember(ctx, t, epc)
+				}
+			}
+		})
+	}
+}
+
+func addMember(ctx context.Context, t *testing.T, epc *e2e.EtcdProcessCluster) {
+	_, err := epc.StartNewProc(ctx, nil, t, e2e.MemberAdd)
+	require.NoError(t, err)
+	require.NoError(t, epc.Procs[len(epc.Procs)-1].Etcdctl().Health(ctx))
+}
+
+func addMemberAsLearner(ctx context.Context, t *testing.T, epc *e2e.EtcdProcessCluster) {
```

Comment:
Please separate function to promote the learner or rename it to `addMemberAsLearner` and promote.

Reply:
ack. updating